### ISSUE TITLE
Inconsistency between binder property paths

### ISSUE THREAD

**Issue by copitz** [spring-projects/spring-boot#17615]
**TL;DR:**
The keys of nested properties bound to a `Map<String, String>` look like `"foo.0.bar"` but should be `"foo[0].bar"`.

**Further explanation:**
This one took me hours to figure out and I managed to work around it by replacing all `"\\.([0-9]+)(\\.|$)"`with `[$1]$2` in the keys of the map - anyway I think that it could help someone stumbling upon this and especially when it could be fixed:

To enable polymorph configuration property bindings (and also retrieve such bindings from other sources like a database or webservice) I let spring boot bind (nested) properties under a certain path to a `Map<String, String>` and use this as `MapConfigurationPropertySource` for the new Binder API to bind them to the concrete classes.

This worked fine until I stumbled upon collections/arrays leading to weired behaviors of the binder (e.g. keeping dots in the property path). After a while I found out that this was caused by the wrong form of the property paths.

Please let me know if I should add some code / test case to make this clearer.

**Comment by mbhave** [comment]
@copitz I'm not sure I fully understood the issue. A minimal sample that showcases the issue or a test case would be very useful.

**Comment by copitz** [comment]
@mbhave Sure, hope this helps: https://github.com/copitz/spring-boot-issue-17615

**Comment by mbhave** [comment]
Thanks for the sample @copitz. I took a look and this is the expected behavior:

When `ApplicationBinderTest` runs, the first binding that occurs is binding properties under the `acme` prefix to `ExampleProperties`. Properties are bound to a `Map<String, Map<String, String>>`. When the binder binds properties to a `Map`, it removes everything that is not an alpha-numeric character. In this case, `[` and `]` get removed and the values in the `Map` end up being `entity-filters.products.0.pattern` and `entity-filters.products.0.include`. This is documented [here](https://docs.spring.io/spring-boot/docs/current/reference/htmlsingle/#boot-features-external-config-relaxed-binding)

> When binding to Map properties, if the key contains anything other than lowercase alpha-numeric characters or -, you need to use the bracket notation so that the original value is preserved. If the key is not surrounded by [], any characters that are not alpha-numeric or - are removed.

This is why it works when you add the `[ ]` for the second test.

When you use Spring Boot's `@ConfigurationProperties` to bind directly to `StoreApplication`, it has the key in the original form (since you are directly binding keys under `acme.applications.foo` ).  

One way to avoid manually fixing the property names would be to specify the properties using the convention mentioned in the docs linked above. Using `acme.applications.foo.[entity-filters.products[0].pattern]` would work as the original `[0]` would be preserved.

We have an open issue #13404 for investigating if map keys can preserve their original value without having to surround them with `[ ]`. I'll close this one in favor of that.

**Comment by copitz** [comment]
Thank you @mbhave for the clarification
